### PR TITLE
[ML] Stochastic derivatives for tree training

### PR DESCRIPTION
**Pull request by tveasey**
The primary change is to use a random downsample (bag) of data for each new tree we train. I extended initialisation to perform a line search over the downsampling ratio to find a good range to search during fine tuning. This is primarily intended to improve runtime on larger data sets, but incidentally improved QoR for many data sets since it tends to decorrelates the errors between trees.

This also tweaks the initial range used for line search to base its start **and end** point on percentile values for the gain and total curvature of a tree. Finally, since I needed to update progress monitoring to account for the extra line searches, I switched to updating progress after each tree which is trained. This gives much finer grained progress monitoring.

### CODE REVIEW

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -665,19 +662,21 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
     //  4. Update predictions and loss derivatives
 
     double eta{m_Eta};
-    double oneMinusBias{eta};
+    std::size_t nextTreeCountToRefreshSplits{
+        forest.size() + static_cast<std::size_t>(std::max(0.5 / eta, 1.0))};
 
-    TDoubleVecVec candidateSplits(this->candidateSplits(frame, trainingRowMask));
+    auto downsampledRowMask = this->downsample(trainingRowMask);
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: Did you try to use your implementation of stratified sampling for regression/classification here? I would expect, it should lead to better results. My intuiting is that using stratified sampling instead of simple sampling would better preserver the data distribution properties.

File: lib/maths/CBoostedTreeImpl.cc [hunk 2]
```diff
@@ -686,28 +685,53 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
...
-        } else if (oneMinusBias < 1.0) {
-            scopeMemoryUsage.add(tree);
-            this->refreshPredictionsAndLossDerivatives(frame, trainingRowMask, 1.0, tree);
-            oneMinusBias = 1.0;
-            forest.push_back(std::move(tree));
+            m_TrainingProgress.increment();
+        } else {
+            candidateSplits = this->candidateSplits(frame, downsampledRowMask);
+            nextTreeCountToRefreshSplits +=
+                static_cast<std::size_t>(std::max(0.5 / eta, 2.0));
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: It would be nice to have a comment explaining what is going on here.

File: lib/maths/CBoostedTreeImpl.cc [hunk 3]
```diff
@@ -686,28 +685,53 @@ CBoostedTreeImpl::trainForest(core::CDataFrame& frame,
...
         }
-        LOG_TRACE(<< "bias = " << (1.0 - oneMinusBias));
 
-        if (m_Loss->isCurvatureConstant() == false) {
-            candidateSplits = this->candidateSplits(frame, trainingRowMask);
+        downsampledRowMask = this->downsample(trainingRowMask);
+
+        if (forest.size() == nextTreeCountToRefreshSplits) {
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: I would be nice to have a comment why recomputing candidate splits is not needed in every iteration.

File: lib/maths/CBoostedTreeFactory.cc [hunk 4]
```diff
@@ -326,70 +350,117 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) {
...
 
-    double gainPerNode;
-    double totalCurvaturePerNode;
-    std::tie(gainPerNode, totalCurvaturePerNode) =
-        this->estimateTreeGainAndCurvature(frame, allTrainingRowsMask);
-    LOG_TRACE(<< "gain per node = " << gainPerNode
-              << " total curvature per node = " << totalCurvaturePerNode);
+    double gainPerNode1stPercentile{gainAndTotalCurvaturePerNode[0].first};
+    double gainPerNode50thPercentile{gainAndTotalCurvaturePerNode[1].first};
+    double gainPerNode90thPercentile{gainAndTotalCurvaturePerNode[2].first};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: So this was partly an empirical observation, in fact I started with a symmetric interval. At 90th percentile we severely curtail the tree, i.e. in all the data sets I tested this majorly under fit, so using a higher value seemed unnecessary. By contrast, we could sometimes still eek more performance out if I cut this off too high. My hand wavy argument would be that since this applies to the base learner we should be overfitting somewhat, so we should prefer to explore smaller values.

File: lib/maths/CBoostedTreeImpl.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: This is something I haven't explored yet. There is some prior art for stochastic gradient boosting, i.e. Friedman, which uses pure random sampling. Also, this approach can be implemented directly on the mask and doesn't require visiting the data, which is attractive from a run time perspective. As discussed offline, I'm going to raise an issue to explore knock on changes we should explore given this change and I'll mention this one.

File: lib/maths/CBoostedTreeFactory.cc [hunk 5]
```diff
@@ -306,14 +325,19 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) {
...
-            static_cast<std::size_t>(2.0 / m_TreeImpl->m_Eta + 0.5);
+        m_TreeImpl->m_MaximumNumberTrees = computeMaximumNumberTrees(m_TreeImpl->m_Eta);
     }
 
     if (m_TreeImpl->m_FeatureBagFractionOverride != boost::none) {
         m_TreeImpl->m_FeatureBagFraction = *(m_TreeImpl->m_FeatureBagFractionOverride);
     }
 
+    double numberFeatures{static_cast<double>(m_TreeImpl->m_Encoder->numberEncodedColumns())};
+    double downSampleFactor{200.0 * numberFeatures /
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Ok so this is derived from the usual rule of thumb that the number of examples for linear regression should be proportional to the number of regressors to ensure independence of *empirical* residuals. I would normally say I need 30 * number of regressors to be confident this condition holds, so this represents a large comfort margin. However, since 1) we have a more expressive model, 2) we use this to upper bound the range of downsampling factors we search using a higher value seems appropriate. On the empirical side, for the data sets I tested we didn't choose values higher than this in the line search (again with some margin), but this could do with more testing. Perhaps the best thing to do would be to expose this parameter in the API and we should also do a fuller study down-the-line.

[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeFactory.cc]: I think 200 is a good number. Funny, I've heard about the rule of thumb with 30x in the context of neural networks years ago. Could never confirm where it originated from :smile: Could you please extract it as a constant with a meaningful name? I wouldn't mind to keep it hidden from users, as it is very technical, but we would need to monitor it on real data.

File: lib/maths/CBoostedTreeFactory.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeFactory.cc]: Ok. I am satisfied with your hand wavy argument :)

File: lib/maths/CBoostedTreeFactory.cc [hunk 6]
```diff
@@ -459,64 +546,130 @@ void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDa
...
+            return tree.m_DownsampleFactor * tree.m_TrainingRowMasks[0].manhattan() > 10.0;
+        };
+
+        m_LogDownsampleFactorSearchInterval =
+            initial + this->testLossLineSearch(frame, applyDownsampleFactorStep,
+                                               std::log(MIN_DOWNSAMPLE_FACTOR_SCALE),
+                                               std::log(MAX_DOWNSAMPLE_FACTOR_SCALE), stepSize)
+                          .value_or(fallback - initial);
+
+        // Truncate the scale to be less than or equal to 1.0.
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: See [d9bf645](https://github.com/elastic/ml-cpp/pull/811/commits/d9bf645250ad160a1a156aec6e91186a113d76df)

File: lib/maths/CBoostedTreeFactory.cc [hunk 7]
```diff
@@ -326,70 +350,117 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) {
...
+    double totalCurvaturePerNode1stPercentile{gainAndTotalCurvaturePerNode[0].second};
+    double totalCurvaturePerNode90thPercentile{gainAndTotalCurvaturePerNode[2].second};
 
+    // Make sure all line search intervals are not empty.
+    gainPerNode1stPercentile = std::min(gainPerNode1stPercentile, 0.1 * gainPerNode90thPercentile);
+    totalCurvaturePerNode1stPercentile = std::min(
+        totalCurvaturePerNode1stPercentile, 0.1 * totalCurvaturePerNode90thPercentile);
+
+    // Search for depth limit at which the tree starts to overfit.
+    if (m_TreeImpl->m_RegularizationOverride.softTreeDepthLimit() == boost::none) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: Thanks!

File: lib/maths/CBoostedTreeImpl.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: I added a catch all comment for amortising cost computing splits. See [56f32ea](https://github.com/elastic/ml-cpp/pull/811/commits/56f32eae2a4f4baa7867dfbd164c389022288cfe). I didn't explicitly mention why I tied to eta, i.e. captures something about how different we expect splits chosen from round-to-round to be. Does this seem sufficient to you?

File: lib/maths/CBoostedTreeImpl.cc (hunk 3 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: See [56f32ea](https://github.com/elastic/ml-cpp/pull/811/commits/56f32eae2a4f4baa7867dfbd164c389022288cfe).

File: lib/maths/CBoostedTreeFactory.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: I kept the default but exposed the parameter as a get out-of-jail of this is very wrong. See [5b9a04a](https://github.com/elastic/ml-cpp/pull/811/commits/5b9a04aafe6d1c30ce7ad3b31df23b2cd82a92f8).

**Comment by tveasey** [comment]
Thanks for the review @valeriy42 and good catch on missing persistence changes! I think I've now addressed all your comments. Can you take another look?

File: lib/maths/CBoostedTreeFactory.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (Winterflower) [reply lib/maths/CBoostedTreeFactory.cc]: Curious: What were the characteristics of the datasets that were used to derive this empirical observation? Would it be possible to take a look at these?
